webglframework.js

```diff
@@ -1,523 +1,480 @@
     for(i in transparent){
-      W.r(transparent[i]);
+      W.render(transparent[i], dt);
     }
     
     // Disable alpha blending for next frame
-    gl.disable(gl.BLEND);
+    W.gl.disable(3042 /* BLEND */);
   },
   
   // Render an object
-  r: (s, center = [0,0,0], vertices, texCoords, buffer) => {
+  render: (object, dt, buffer) => {
 
     // If the object has a texture
-    if (s.b.id) {
+    if (object.b.id) {
 
       // Set the texture's target (2D or cubemap)
-      // gl.bindTexture(gl.TEXTURE_2D, W.textures[s.b.id]);
-      gl.bindTexture(3553, W.textures[s.b.id]);
+      W.gl.bindTexture(3553 /* TEXTURE_2D */, W.textures[object.b.id]);
 
       // Pass texture 0 to the sampler
-      gl.uniform1i(gl.getUniformLocation(W.P, 'sampler'), 0);
+      W.gl.uniform1i(W.gl.getUniformLocation(W.program, 'sampler'), 0);
     }
 
     // If the object has a transition, increment its timer...
-    if(s.f < s.t) s.f += W.dt;
-    // ...but don't let it go over the transition duration.
-    if(s.f > s.t) s.f = s.t;
+    if(object.f < object.t) object.f += dt;
     
-    //s.vertices = vertices;
-    //s.texCoords = texCoords;
-    s.center = center;
+    // ...but don't let it go over the transition duration.
+    if(object.f > object.t) object.f = object.t;
 
-    // Set the object as the currently updated object.
-    W.N = s.n;
+    // Compose the model matrix from lerped transformations
+    W.next[object.n].m = W.transition(object.n);
 
-    // Compose the model matrix from lerped transformations.
-    W.n[s.n].m = W.t();
+    // If the object is in a group:
+    if(W.next[object.g]){
 
-    // If the object is in a group…
-    if (W.n[s.g]) {
-      // …left-multiply the model matrix by the group's model matrix.
-      W.n[s.n].m.preMultiplySelf(W.n[s.g].m);
+      // premultiply the model matrix by the group's model matrix.
+      W.next[object.n].m.preMultiplySelf(W.next[object.g].m);
     }
 
-    gl.uniformMatrix4fv(  // send it to the shaders
-      gl.getUniformLocation(W.P, 'm'),
+    // send the model matrix to the vertex shader
+    W.gl.uniformMatrix4fv(
+      W.gl.getUniformLocation(W.program, 'm'),
       false,
-      W.n[s.n].m.toFloat32Array()
+      W.next[object.n].m.toFloat32Array()
     );
     
-    // Ignore camera, light, groups
-    if(!["C","L"].includes(s.n) && s.T != "g"){
-      s.center = center;
-
-      // Set the position buffer
-      
-      //gl.bindBuffer(gl.ARRAY_BUFFER, gl.createBuffer());
-      gl.bindBuffer(34962, W.vertices[s.T]);      
-      
-      // gl.vertexAttribPointer(buffer = gl.getAttribLocation(W.P, 'position'), 3, gl.FLOAT, false, 0, 0)
-      gl.vertexAttribPointer(buffer = gl.getAttribLocation(W.P, 'position'), 3, 5126, false, 0, 0)
+    // Don't render inisible items (camera, light, groups)
+    if(!['camera','light','group'].includes(object.type)){
       
-      gl.enableVertexAttribArray(buffer);
+      // Send the position buffer to the vertex shader
```

Comment:
This doesn't really send anything to the shader. The buffer's data has already been uploaded to the GPU earlier. The code here sets up the shader's attributes to point to the right buffers and tells them what data types are stored in these buffers.